### PR TITLE
Refactor ExecutionFrame and SessionState to reduce memory allocations and improve data locality

### PR DESCRIPTION
**Pull request by yuslepukhin**
**Description**: Describe your changes.
Reduce per-Run() memory allocations in SessionState, ExecutionFrame and Patterns.

**Motivation and Context**
This changes based on customer request to reduce number of memory allocations.

**Before**

BeforeRefactoring 2022-06-06 173710

**After**

After Refactor Screenshot 2022-06-08 115058

### CODE REVIEW

File: onnxruntime/core/framework/execution_frame.cc [hunk 1]
```diff
@@ -362,12 +362,12 @@
...
-      int ort_value_idx = fetch_mlvalue_idxs[idx];
-
-      auto custom_alloc_entry = fetch_allocators.find(idx);
-      if (custom_alloc_entry != fetch_allocators.cend()) {
-        custom_allocators_[ort_value_idx] = custom_alloc_entry->second;
+    custom_allocators_.reserve(fetch_allocators.size());
+    const auto idx_size = fetch_mlvalue_idxs.size();
+    for (const auto& e : fetch_allocators) {
+      if (e.first < idx_size) {
+        int ort_value_idx = fetch_mlvalue_idxs[e.first];
```
Comment (yuslepukhin) [inline onnxruntime/core/framework/execution_frame.cc:369]: > in [](http://example.com/codeflow?start=7&length=3)

size_t or auto

File: onnxruntime/core/framework/session_state.h [hunk 2]
```diff
@@ -206,19 +206,31 @@
   /**
   Get cached memory pattern based on input shapes
   Must be called only when all values contain tensors
+  In training scenarios, the cache may be updated so
+  the callers would receive a copy of inferred shapes
+  made under mutes being held. In inference scenarios,
```
Comment (yuslepukhin) [inline onnxruntime/core/framework/session_state.h:211]: >mutes [](http://example.com/codeflow?start=13&length=5)

mutex

File: onnxruntime/core/framework/session_state.cc [hunk 3]
```diff
@@ -663,36 +670,65 @@
...
   return Status::OK();
 }
-#endif
 
-const MemoryPatternGroup* SessionState::GetMemoryPatternGroup(const gsl::span<const OrtValue>& tensor_inputs,
-                                                              const std::vector<int>& feed_mlvalue_idxs,
-                                                              std::unordered_map<int, TensorShape>& inferred_shapes) const {
+// XXX: it is not clear how the mutex protects MemoryPatternGroup
+// since the pointer is returned. Execution frame is reading it during construction
+// but other inference/training Run() update it via UpdateMemoryPatternGroupCache() calls
```
Comment (yuslepukhin) [inline onnxruntime/core/framework/session_state.cc:676]: Change the comment to provide explanation

File: onnxruntime/core/framework/simple_tensor_allocator.h [hunk 4]
```diff
@@ -19,17 +20,17 @@
 
  public:
   SimpleTensorAllocator(const ExecutionPlanBase& execution_plan, const SessionState& session_state,
-                        std::vector<BufferUniquePtr>& /*weights_buffers*/)
+                        InlinedVector<BufferUniquePtr>& /*weights_buffers*/)
       : ITensorAllocator(session_state),
         seq_plan_(execution_plan) {}
 
-  common::Status FinalizePlan(std::unordered_map<std::string, size_t>& planned_memory_sizes_in_byte) override {
```
Comment (yuslepukhin) [inline onnxruntime/core/framework/simple_tensor_allocator.h:26]: >planned_memory_sizes_in_byte [](http://example.com/codeflow?start=71&length=28)

comment out as unused

File: onnxruntime/core/framework/session_state.h [hunk 5]
```diff
@@ -480,10 +485,12 @@ class SessionState {
 
   // lock for the mem_patterns_
   mutable OrtMutex mem_patterns_lock_;
-
   // cache for the generated mem_patterns. key is calculated based on input shapes.
-  mutable std::map<int64_t, std::unique_ptr<MemoryPatternGroup>> mem_patterns_;
-  mutable std::map<int64_t, std::unordered_map<int, TensorShape>> shape_patterns_;
+  // must be a node based container as a pointer is cached.
+  mutable NodeHashMap<int64_t, MemoryPatternGroup> mem_patterns_;
+  // This is mutable under mutex in training scenarios so execution frame would make a copy
```
[the remark this answers is not in the document]
Comment (yuslepukhin) [reply onnxruntime/core/framework/session_state.h:491]: The second one can be non-mutable, but the first one, can still be updated if entry is not present

File: onnxruntime/core/framework/session_state.cc [hunk 6]
```diff
@@ -191,11 +191,16 @@ Status SessionState::CreateKernels(const KernelRegistryManager& kernel_registry_
       ORT_RETURN_IF_ERROR(kernel_registry_manager.CreateKernel(node, exec_provider, *this, kci, session_kernels_[node.Index()]));
     }
   }
-  node_index_info_ = std::make_unique<NodeIndexInfo>(*graph_viewer_, ort_value_name_idx_map_);
+  node_index_info_.emplace(*graph_viewer_, ort_value_name_idx_map_);
   return Status::OK();
 }
 
-const SequentialExecutionPlan* SessionState::GetExecutionPlan() const { return p_seq_exec_plan_.get(); }
+const SequentialExecutionPlan* SessionState::GetExecutionPlan() const {
```
[the remark this answers is not in the document]
Comment (yuslepukhin) [reply onnxruntime/core/framework/session_state.cc:198]: We can not return optional, since it would require a return by value and a copy.

**Review by pranavsharma — APPROVED** [review]

File: include/onnxruntime/core/framework/ortmemoryinfo.h [hunk 7]
```diff
@@ -38,6 +38,20 @@ struct OrtMemoryInfo {
...
+  // This is to make OrtMemoryInfo a valid key in hash tables
+  // we ignore device id
+  size_t Hash() const {
+    auto h = std::hash<int>()(alloc_type);
+    HashCombine(std::hash<int>()(mem_type), h);
+    HashCombine(std::hash<int>()(id), h);
+    HashCombine(std::hash<const char*>()(name), h);
+    return h;
+  }
+
```
Comment (skottmckay) [inline include/onnxruntime/core/framework/ortmemoryinfo.h:54]: In most use cases the name doesn't matter. What does is the mem_type and device (which duplicates the id value). 

Should we have an alternative hash implementation that allows using that info in the hash so this knowledge is in one place? That could be used for SessionState::allocators_.

Comment (yuslepukhin) [reply include/onnxruntime/core/framework/ortmemoryinfo.h:54]: I saw that alternate comparison function. After thinking a bit, I decided to make it equivalent of operator< here in the class. The way I see it:
1) We have special hash for allocators_ just like a special operator<
2) Or we change both of them here and use everywhere.

Comment (skottmckay) [reply include/onnxruntime/core/framework/ortmemoryinfo.h:54]: operator< seems flawed. the OrtDevice is arguably the most important thing but is ignored by that implementation. 

it's fine to leave the allocator related hashing separate. it's a larger design discussion to rationalize the whole allocator setup and probably falls under that to decide what's best long term.

Comment (iK1D) [reply include/onnxruntime/core/framework/ortmemoryinfo.h:54]: As stated in this [link](http://en.cppreference.com/w/cpp/utility/hash):

> There is no specialization for C strings. std::hash produces a hash of the value of the pointer (the memory address), it does not examine the contents of any character array.

So is the std::hash<const char*> here an expected one?

**Pull request merged**